### PR TITLE
Settings redesign

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
In continuation of #1066

### CODE REVIEW

**Comment by lognaturel** [comment]
Looking great! I hope merging the conflicts isn't too time-consuming.

Two things I noticed are missing: the actual splash screen image and the form metadata.

**Comment by shobhitagarwal1612** [comment]
@lognaturel Done

**Comment by lognaturel** [comment]
Looking good! 

Fonts are pretty small. Are those the Material design standards? Could we just bump everything up by a couple of sp? I changed the font size to large at the device level and that feels very comfortable to me. Maybe those sizes could be the defaults? I'm willing to be talked out of this if you think sticking to standards and helping users change the font size at the device level is better.

Passwords should have a show/hide option. We can also file this as a separate issue and do it as the eye overlay which I think is now standard. Or perhaps dialogs should be handled separately. I notice the color of the dialog buttons are also inconsistent.

Speaking of color inconsistency! I sometimes get a mix of blue and teal on Nexus 5, API 22. Everything is consistently blue on my Moto G4, Android 7.0.0.


I haven't done a proper review of the code yet but one quick thing I noticed is that there is now only one key for the auto send setting. That's probably the right thing to do but the reason @yanokwa mentioned using the same preference keys is for backwards compatibility. We really don't want users not getting their forms as expected because autosend was turned off without them noticing. Also, if users have the separate settings as part of configuration QR codes, those should still work. You might want to look at `FormMetadataMigrator` for ideas on how to handle that.

**Comment by shobhitagarwal1612** [comment]
@lognaturel : 
> Fonts are pretty small. Are those the Material design standards? Could we just bump everything up by a couple of sp? I changed the font size to large at the device level and that feels very comfortable to me. Maybe those sizes could be the defaults? I'm willing to be talked out of this if you think sticking to standards and helping users change the font size at the device level is better.

The font size of the entire app or just the settings screen?


> I haven't done a proper review of the code yet but one quick thing I noticed is that there is now only one key for the auto send setting. That's probably the right thing to do but the reason @yanokwa mentioned using the same preference keys is for backwards compatibility. We really don't want users not getting their forms as expected because autosend was turned off without them noticing. Also, if users have the separate settings as part of configuration QR codes, those should still work. You might want to look at FormMetadataMigrator for ideas on how to handle that.

Added backward compatibilty for importing from QRCode

**Comment by lognaturel** [comment]
For the font size, I'm talking specifically about the settings screen for now.

File: collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java [hunk 1]
```diff
@@ -685,6 +688,41 @@ private boolean loadSharedPreferencesFromFile(File src) {
         return res;
     }
 
+    /**
+     * This method is to provide backward compatibility with v1.7.0 and below
+     * Autosend was originally set into separate wifi and cellular autosend settings
+     */
+    private void checkForBackwardCompatility(Map<String, ?> entries, Editor prefEdit) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java]: A method name that starts with `check` implies it's going to return a boolean. I would name this something like `migrateAutosendPrefKeys`.

File: collect_app/src/main/java/org/odk/collect/android/preferences/DisabledPreferencesRemover.java [hunk 2]
```diff
@@ -66,32 +42,86 @@
...
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks whether the preferenceGroup actually has any child preferences defined
+     */
+    private boolean hasChildPrefs(String preferenceKey) {
+        String[] preferenceScreensWithNoChildren = {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/DisabledPreferencesRemover.java:127]: Isn't this the list WITH children, not with NO children?

**Review by lognaturel — CHANGES_REQUESTED** [review]
I don't see any code for doing the migration from two autosend preferences to one in the case of a normal update. To verify:
1. `git checkout v1.7.0` and install to device
1. check both "autosend with wifi" and "autosend with network"
1. checkout this branch and install to device
Now I should see "Wifi or cellular" selected for "Autosend" in the form management settings but I don't.

Another small thing I noticed is that if I uncheck everything in Admin Settings > User Settings, the Server category disappears but the other three categories are still there. Form management is clickable but empty.

Other than that, it looks ready to merge! The first one is a must-fix. The second we can file as a separate issue.

File: collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java (hunk 1 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java]: Fixed

File: collect_app/src/main/java/org/odk/collect/android/preferences/DisabledPreferencesRemover.java (hunk 2 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/preferences/DisabledPreferencesRemover.java:127]: Fixed

**Comment by shobhitagarwal1612** [comment]
> I don't see any code for doing the migration from two autosend preferences to one in the case of a normal update. To verify:
> 
> git checkout v1.7.0 and install to device
> check both "autosend with wifi" and "autosend with network"
> checkout this branch and install to device
> Now I should see "Wifi or cellular" selected for "Autosend" in the form management settings but I don't.

Done

**Comment by shobhitagarwal1612** [comment]
@lognaturel : 

> Another small thing I noticed is that if I uncheck everything in Admin Settings > User Settings, the Server category disappears but the other three categories are still there. Form management is clickable but empty.

I have fixed this. Although there is still a minute bug left. The form management is visible and clickable if all the preferences are hidden but it works if any one or more other preferences is/are unhidden. We can file a separate issue for fixing this.

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
I verified that when I check check both auto send settings in v1.7.0, install this branch, then auto send is set to Wifi & Cellular as it should be.

However, when I checked auto send with network in v1.7.0, exported the settings file (Admin settings > menu > Save Settings to Disk), moved the generated file to /sdcard/odk/collect.settings, installed this branch and then launched it, I saw the "settings successfully imported" toast but auto send was still set to "Off" (It should have said "cellular"). This will likely be addressed by #1133 but I also wanted to keep a record of it here.

**Pull request merged**